---
title: "Mitochondrial tRNA counts"
author: "Laura White"
date: "2023-12-26"
output: html_document
---
```{r global options, include = FALSE}
knitr::opts_chunk$set(message=FALSE)
library(tidyverse)
library(colorblindr)
library(cowplot)
library(here)
here::i_am("Rmd/mitocounts.Rmd")
```
We want to know if our mitochondrial tRNA aligning reads are genuine.
First, let's examine the abundance of different tRNA species aligned to different genome references.
```{r}
read_tRNA <- function(x, sample, ref) {
  df <- readr::read_tsv(x, col_names = c("chrom", "start", "pos", "score"), show_col_types = FALSE)
  
  df <- df %>%
    dplyr::mutate(sample = sample, ref = ref)
  
  return(df)
}

# yeast aligned to nuclear tRNA reference only
read_tRNA(here("rebasecalling/bedgraphs/allreads/WTyeast004_20231111_1104_P2S-00519-B_PAQ47538_fa3726ec.rna004_130bps_sup@v3.0.1.bg"), "WT", "nuclear") -> nuc_WT
read_tRNA(here("rebasecalling/bedgraphs/allreads/Pus4yeast004_20231208_1152_MN42516_FAX71838_e2d70aac.rna004_130bps_sup@v3.0.1.bg"), "pus4del", "nuclear") -> nuc_pus4

# yeast with mito and nuclear tRNA reference
read_tRNA(here("rebasecalling/bedgraphs/mito_analysis/WTyeast004_20231111_1104_P2S-00519-B_PAQ47538_fa3726ec.rna004_130bps_sup@v3.0.1.bg"), "WT", "both") -> all_WT
read_tRNA(here("rebasecalling/bedgraphs/mito_analysis/Pus4yeast004_20231208_1152_MN42516_FAX71838_e2d70aac.rna004_130bps_sup@v3.0.1.bg"), "pus4del", "both") -> all_pus4

combined <- list(nuc_WT, nuc_pus4, all_WT, all_pus4)

bind_rows(combined) %>%
  mutate(pos = as.numeric(pos)) %>%
  mutate(score = as.numeric(score)) %>%
  separate(chrom, c("origin", "tRNA", "AA", "anticodon", "family", "isotype")) %>%
  select(-tRNA) %>%
  mutate(anticodon = str_replace_all(anticodon, "U", "T")) -> tRNAcov # u and t inconsistency across refs

# this produces some NAs for t thermophila isotypes

rm(nuc_WT, nuc_pus4, all_WT, all_pus4, combined)
```

The above gives us a tidy tibble with coverage x position for each tRNA in each sample.
I then make the assumption that the maximum coverage score for each gene is a better proxy coverage value than (for example) the mean coverage on that reference, because per-nucleotide coverage is being depressed by Nanopore error rates and mismatches/indels due to tRNA modifications.
```{r}
tRNAcov %>%
  group_by(sample, ref, origin, AA, anticodon, family, isotype) %>%
  slice_max(score, with_ties = F) %>%
  ungroup() %>%
  unite(tRNA, origin, AA, anticodon, sep = "-") %>%
  select(-pos) -> rollup
```

The code below adds up max scores for family & gene number info so that we have summary scores for each isodecoder.
```{r}
rollup %>%
  separate(tRNA, c("origin", "AA", "anticodon"), extra = "drop") %>%
  group_by(origin, AA, anticodon, sample, ref) %>%
  summarize(score = sum(score)) %>%
  unite(tRNA, origin, AA, anticodon, sep = "-") %>%
  ungroup() %>%
  group_by(sample, ref) %>%
  mutate(total_score = sum(score)) %>%
  mutate(CPM = score/total_score*1000000) -> isodecoder_scores
```

So one thing we want to ask is do the abundance scores for nuclear tRNAs change when we add the mitochondrial tRNAs to the reference.
Quick plot of everything aligned to the two references.
```{r}
isodecoder_scores %>%
#  filter(str_starts(tRNA, "Gly") | str_starts(tRNA, "Pro")) %>%
  ggplot(., aes(y = CPM, x = tRNA, fill = fct_rev(sample))) +
  geom_bar(stat = "identity", width=.5, position = "dodge") +
  scale_fill_OkabeIto(order = c(2,1)) +
  theme_cowplot() +
  theme(legend.position = 'right',
    legend.title = element_blank(),
    legend.direction ='vertical') +
  theme(axis.text.x = element_text(angle = 90, vjust = .35)) +
  theme(aspect.ratio=1/10) +
  labs(title = "",
       x = "",
       y = "Counts / Million") +
  scale_y_log10() +
  facet_wrap(~ref, ncol =1)
```
Are these abundance values consistent with what we'd expect to see for mt-tRNAs? I'd expect them to be 1-5% abundance at best of their cytoplasmic counterparts.
```{r}
isodecoder_scores %>%
  filter(ref == "both") %>%
  separate(tRNA, c("origin", "tRNA"), extra = "merge") %>%
  group_by(origin, sample) %>%
  summarize(mean_CPM = mean(CPM, na.rm = TRUE)) -> mean_cpm

total_cpm <- mean_cpm %>%
  group_by(sample) %>%
  summarize(total_CPM = sum(mean_CPM))

mean_cpm <- mean_cpm %>%
  left_join(total_cpm, by = "sample")

mean_cpm <- mean_cpm %>%
  mutate(percent_CPM = (mean_CPM / total_CPM) * 100)


# put WT first
mean_cpm$sample <- fct_rev(mean_cpm$sample)

ggplot(mean_cpm, aes(x = origin, y = mean_CPM, fill = origin)) +
  geom_bar(stat = "identity") +
  theme_minimal() +
  labs(title = "Mean tRNA counts per million,\nnuclear encoded vs. mt-tRNAs",
       x = "",
       y = "Mean CPM") +
  theme(legend.position = "none") +
  theme_cowplot() +
  scale_fill_OkabeIto() +
  facet_wrap(~sample)
```

Then let's look at fold change comparison on the nuclear tRNAs only between these two reference alignments.
```{r}
isodecoder_scores %>%
  separate(tRNA, c("origin", "tRNA"), extra = "merge") %>%
  filter(origin == "nuc" & sample == "WT") %>%
  select(-score, -total_score) %>%
  pivot_wider(names_from = ref, values_from = CPM, values_fill = 0) %>%
  mutate(foldc = ((both/nuclear)-1)) -> fc

fc %>%
  ggplot(., aes(y = foldc, x = tRNA)) +
  geom_bar(stat = "identity", width=.5, position = "dodge") +
  scale_fill_OkabeIto(order = c(3,7)) +
  theme_minimal() +
  theme(legend.position = 'right',
    legend.title = element_blank(),
    legend.direction ='vertical') +
  theme(axis.text.x = element_text(angle = 90, vjust = .35)) +
  theme(aspect.ratio=1/4) +
  labs(title = "Nuclear tRNA foldchange when mt-tRNAs are added to reference",
       x = "",
       y = "Fold Change")
```
And for my interest, foldchange between Pus4 and WT when aligned to the reference with both mito and nuc tRNAs?
```{r}
isodecoder_scores %>%
  filter(ref == "both") %>%
  select(-score, -total_score) %>%
  pivot_wider(names_from = sample, values_from = CPM, values_fill = 0) %>%
  mutate(foldc = ((pus4del/WT)-1)) -> fc

fc %>%
  ggplot(., aes(y = foldc, x = tRNA)) +
  geom_bar(stat = "identity", width=.5, position = "dodge") +
  scale_fill_OkabeIto(order = c(3,7)) +
  theme_cowplot() +
  theme(legend.position = 'right',
    legend.title = element_blank(),
    legend.direction ='vertical') +
  theme(axis.text.x = element_text(angle = 90, vjust = .35)) +
  theme(aspect.ratio=1/8) +
  labs(title = "tRNA foldchange on Pus4 deletion",
       x = "",
       y = "Fold Change")
```

One more question. How do the MAPQ score distributions change as we add the mito-tRNAs to the mix?
Here I've got just the RNA004 wild type yeast samples.
```{r}
nuconly_MAPQ <- read_tsv(here("metadata/WT_nuconly_mapq_scores.txt"), col_names = "MAPQ") %>%
  mutate(Sample = "Nuconly")

withmito_MAPQ <- read_tsv(here("metadata/WT_withmito_mapq_scores.txt"), col_names = "MAPQ") %>%
  mutate(Sample = "Withmito")

combined_MAPQ <- bind_rows(nuconly_MAPQ, withmito_MAPQ)

mean_mapq_per_sample <- combined_MAPQ %>%
  group_by(Sample) %>%
  summarize(mean_MAPQ = mean(MAPQ, na.rm = TRUE))

library(scales)  # For formatting labels

# Function to create labels
log_breaks <- function(n = 5) {
  res <- extended_breaks(n = n)(range(combined_MAPQ$MAPQ))
  return(res)
}

# Custom labels for the breaks
log_labels <- function(x) {
  return(as.character(round(x, 0)))
}

ggplot(combined_MAPQ, aes(x = MAPQ, fill = Sample)) +
  geom_density(alpha = .4) +
  labs(title = "Alignment mapping scores improve with \naddition of mt-tRNAs to the reference",
       x = "MAPQ Score",
       y = "Density") +
  theme_cowplot() +
  scale_x_log10(breaks = log_breaks(), labels = log_labels) +
  scale_fill_OkabeIto()
  

```


