---
title: "Modomics - all tRNA"
author: "Laura White"
date: "2024-06-24"
output: html_document
---
```{r}
knitr::opts_chunk$set(echo = F)
library(tidyverse)
library(cowplot)
library(patchwork)
library(stringr)
library(here)

here::i_am("rmd/modomics_alltRNA.Rmd")
```
Here's a function to read in fasta files of tRNA sequences from Modomics database (http://modomics.genesilico.pl/) and parse their headers so that we can make a clean table of tRNA sequences with modifications and their metadata.
```{r}
# to parse a single FASTA file with headers from Modomics.org
parse_fasta <- function(file_path) {
    fasta_lines <- read_lines(file_path)

    # Split lines into headers and sequences
    headers <- fasta_lines[str_detect(fasta_lines, "^>")]
    sequences <- fasta_lines[!str_detect(fasta_lines, "^>")]

    # Function to parse headers
    parse_header <- function(header) {
        key_value_pairs <- str_split(str_remove_all(header, "^>"), "\\|", simplify = FALSE)[[1]]
        key_value_pairs <- lapply(key_value_pairs, function(x) str_split(x, ":", simplify = TRUE)[1, 2])
        return(as.data.frame(t(key_value_pairs), stringsAsFactors = FALSE))
    }

    # Parse all headers
    parsed_headers <- do.call(rbind, lapply(headers, parse_header))

    # Assign column names
    colnames(parsed_headers) <- c("ID", "Name", "SOterm", "Type", "AA", "Anticodon", "Cellular_Localization", "Species")

    # Combine headers and sequences
    final_data <- data.frame(parsed_headers, Sequence = sequences)

    # Tidy up the data
    final_data <- final_data %>%
        mutate(across(everything(), ~str_trim(.))) %>% 
        mutate(across(where(is.character), ~str_remove_all(., pattern = "^[^:]+:"))) %>%
        select(-ID, -SOterm, -Type)

    return(final_data)
}
```
Now let's look at the fasta file containing all tRNA sequences in Modomics. We have 949 entries in this fasta.
```{r}
parse_fasta(here("Modomics/modified_tRNA_all_all_rna_sequences.fasta")) -> all_modified
```

Looking at the whole dataset, which species have the most entries?
```{r}
species_counts <- all_modified %>%
                  group_by(Species) %>%
                  summarise(Count = n())

print(species_counts)

# Assuming you already have the species_counts dataframe
species_counts <- all_modified %>%
                  group_by(Species) %>%
                  summarise(Count = n())

# Filter for species with counts >= 10
species_counts_filtered <- species_counts %>% 
                           filter(Count >= 5)

ggplot(species_counts_filtered, aes(x = reorder(Species, -Count), y = Count)) +
    geom_bar(stat = "identity") +
    theme_minimal() +
    theme(axis.text.x = element_text(angle = 90, hjust = 1)) +
    labs(title = "Organisms with at least 10 tRNAs with annotated modifications in Modomics",
         x = "Species",
         y = "Count")
```

I'd also like to do something looking at the identity / distribution of the modification types across different species.
```{r}

# Modified function to return a list of non-standard characters
extract_non_standard <- function(sequence) {
  str_extract_all(sequence, "[^ACGU]") %>% unlist()
}

# Apply the modified function
all_modified <- all_modified %>%
                mutate(NonStandardChars = sapply(Sequence, extract_non_standard))

# Creating a tidy distribution summary,
# Filter out for characters that are not A, C, G, U, or "any ribonucleoside"
# remove organism = synthetic
distribution_summary <- all_modified %>%
                        unnest(NonStandardChars) %>%
                        count(Species, NonStandardChars) %>%
                        filter(Species != "synthetic construct") %>%
                        filter(NonStandardChars != ""))

modomics <- read_csv(here("Modomics/modomicscodes.csv")) %>%
  select(Name, `Short Name`, `RNAMods code (2023)`) %>%
  rename(modname = Name, shortname = `Short Name`, NonStandardChars = `RNAMods code (2023)`) %>%
  group_by(NonStandardChars) %>%
  arrange(NonStandardChars, shortname) %>%
  # filter duplicate shortnames that are the same base modifications but with 5'-monophosphates
  filter(!(startsWith(shortname, "p") & (NonStandardChars %in% lag(NonStandardChars) | NonStandardChars %in% lead(NonStandardChars))))


annotated <- left_join(distribution_summary, modomics)
annotated %>% distinct(NonStandardChars) -> allspecies

desired_species <- c("Escherichia coli", "Homo sapiens", "Saccharomyces cerevisiae", "Drosophila melanogaster", "Danio rerio", "Tetrahymena thermophila", "Enterobacteria phage T4")
annotated %>% filter(Species %in% desired_species) %>% distinct(NonStandardChars) -> subsetspecies
# After excluding "." (annotated as any ribonucleotide), this gives us 68 unique tRNA mods or 55 in our species of inteest.

left_join(subsetspecies, modomics) -> annotatedsubset
```


```{r}

# Filter data to include only the desired species
filtered_distribution_summary_subset <- distribution_summary %>% filter(Species %in% desired_species) #

# Applying log scale transformation with an offset of 1
filtered_distribution_summary_subset$n_log <- log1p(filtered_distribution_summary_subset$n)

species_counts_filtered_subset <- species_counts_filtered %>%
                                  filter(Species %in% desired_species)

# Order Species based on Count
species_order <- species_counts_filtered_subset %>%
                 arrange(Count) %>%
                 .$Species

filtered_distribution_summary_subset$Species <- factor(filtered_distribution_summary_subset$Species, levels = species_order)
species_counts_filtered_subset$Species <- factor(species_counts_filtered_subset$Species, levels = species_order)

# Create the plots (Heatmap and Bar Plot) using the subset data

# Heatmap Plot
heatmap_plot <- ggplot(filtered_distribution_summary_subset, aes(x = NonStandardChars, y = Species, fill = n_log)) +
    geom_tile() +
    scale_fill_gradient(low = "white", high = "red", name = "Counts of Modifications") +
    theme_minimal() +
    theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5),
          axis.title.x = element_blank(),
          axis.text.y = element_text(size = 7)) +
    labs(y = "Species") +
  theme_cowplot()

# Bar Plot for Total Counts
total_counts_plot <- ggplot(species_counts_filtered_subset, aes(x = Species, y = Count)) +
    geom_bar(stat = "identity", fill = "red") +
    coord_flip() +
    theme_minimal() +
    theme(axis.text.y = element_blank(),
          axis.title.y = element_blank(),
          axis.title.x = element_text(size = 10)) +
    labs(x = "Total Modifications")


# Arrange the plots side by side
combined_plot_a <- total_counts_plot + heatmap_plot + plot_layout(widths = c(1, 10))

# Display the combined plot
print(combined_plot_a)

```
Now, key question: what modifications are not present in the 6 species I'm pretty sure I can get samples for, and in what organisms can I find those modifications?
```{r}
# Assuming distribution_summary contains a column 'NonStandardChars' with nonstandard characters as a string
# Filter data to include only the desired species
desired_species_data <- distribution_summary %>%
                        filter(Species %in% desired_species, !is.na(Species))

# Extract unique nonstandard characters from the desired species
unique_chars_desired <- unique(unlist(strsplit(unlist(desired_species_data$NonStandardChars), "")))

# Extract unique nonstandard characters from the entire dataset
unique_chars_all <- unique(unlist(strsplit(unlist(distribution_summary$NonStandardChars), "")))

# Identify nonstandard characters not in desired species
chars_not_in_desired <- setdiff(unique_chars_all, unique_chars_desired)

# Filter the original distribution_summary for species containing any of the chars_not_in_desired and exclude NAs in Species
species_with_chars <- distribution_summary %>%
                      filter(sapply(strsplit(NonStandardChars, ""), function(char_list) any(chars_not_in_desired %in% char_list)), 
                             !is.na(Species)) %>%
                      .$Species %>%
                      unique()

```
There are only 13 modifications absent in the 6 I've chosen that are present in other organisms. Cool.

So what if we plot those organisms' tRNA modifications?
```{r}
# List of desired species

# Filter data to include only the desired species
filtered_distribution_summary_subset <- distribution_summary %>% filter(Species %in% species_with_chars)

# Applying log scale transformation with an offset of 1
filtered_distribution_summary_subset$n_log <- log1p(filtered_distribution_summary_subset$n)

species_counts_filtered_subset <- species_counts_filtered %>%
                                  filter(Species %in% species_with_chars)

# Order Species based on Count
species_order <- species_counts_filtered_subset %>%
                 arrange(Count) %>%
                 .$Species

filtered_distribution_summary_subset$Species <- factor(filtered_distribution_summary_subset$Species, levels = species_order)
species_counts_filtered_subset$Species <- factor(species_counts_filtered_subset$Species, levels = species_order)

# Create the plots (Heatmap and Bar Plot) using the subset data

# Heatmap Plot
heatmap_plot <- ggplot(filtered_distribution_summary_subset, aes(x = NonStandardChars, y = Species, fill = n_log)) +
    geom_tile() +
    scale_fill_gradient(low = "white", high = "red", name = "Character Count") +
    theme_minimal() +
    theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5),
          axis.title.x = element_blank(),
          axis.text.y = element_text(size = 7)) +
    labs(y = "Species") +
  theme_cowplot()

# Bar Plot for Total Counts
total_counts_plot <- ggplot(species_counts_filtered_subset, aes(x = Species, y = Count)) +
    geom_bar(stat = "identity", fill = "red") +
    coord_flip() +
    theme_minimal() +
    theme(axis.text.y = element_blank(),
          axis.title.y = element_blank(),
          axis.title.x = element_text(size = 10)) +
    labs(x = "Total Count")


# Arrange the plots side by side
combined_plot_b <- total_counts_plot + heatmap_plot + plot_layout(widths = c(1, 5))

# Display the combined plot
combined_plot <- combined_plot_a / combined_plot_b
print(combined_plot)
```
How about looking only at the modifications not present in the subset we think we'll sequence?
```{r}
# Filter the distribution_summary for sequences containing any of the chars_not_in_desired
distribution_summary_filtered <- distribution_summary %>%
                                 filter(sapply(strsplit(NonStandardChars, ""), function(char_list) any(chars_not_in_desired %in% char_list)))

# Create a new dataframe for plotting
plot_data <- distribution_summary_filtered %>%
              count(Species, NonStandardChars) %>%
              filter(NonStandardChars %in% chars_not_in_desired)

# Heatmap Plot
heatmap_plot <- ggplot(plot_data, aes(x = NonStandardChars, y = Species, fill = n)) +
    geom_tile() +
    scale_fill_gradient(low = "white", high = "red", name = "Count") +
    theme_cowplot() +
    theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5),
          axis.title.x = element_blank(),
          axis.text.y = element_text(size = 7)) +
    labs(y = "Species", x = "Non-standard Characters")

# Display the plot
print(heatmap_plot)

```
So that tells me none of them have more than two unique modifications of the 13/69 that are absent in our samples so far. Getting a plant would only give me one new one, and that same modification (which I can't tell the identify of without getting a better lookup system right now) is also present in _S. griseus_. Weirdly, other mammals (rats and cows) seem to have some modifications not characterized in human cells. Overall, it seems like adding _S. griseus_, _H. volcanii_ and possibly _B. taurus_ or _R. norvegicus_ RNA would be worthwhile. Rat's going to be easier to get on campus, I'm sure.

Another thing we want to do is get from these to a bed file like `https://github.com/novoalab/Nano-tRNAseq/blob/c14dcaaf4e59164d5dc43716019b8d39f2d554da/ref/yeast.tRNA.modomics.tsv.bed#L4`
After we do this, we'll need to do is convert the unicode symbols into the Modomics short names to make the modifications more human-readable.

```{r}
generate_bed_data <- function(df) {
  bed_data_list <- lapply(1:nrow(df), function(i) {
    seq_name <- df$Name[i]
    aa <- df$AA[i]
    anticodon <- df$Anticodon[i]
    cellular_location <- df$Cellular_Localization[i]
    species <- df$Species[i]
    sequence <- df$Sequence[i]
    special_chars <- str_locate_all(sequence, "[^A-Za-z]")[[1]]  # Adjust regex for your needs
    if (is.null(special_chars) || nrow(special_chars) == 0) {
      return(NULL)  # No special characters found
    }
    special_chars_df <- as.data.frame(special_chars)
    special_chars_df$Name <- seq_name
    special_chars_df$AA <- aa
    special_chars_df$Anticodon <- anticodon
    special_chars_df$Cellular_Localization <- cellular_location
    special_chars_df$Species <- species
    special_chars_df$ASCII_mod <- str_sub(sequence, special_chars_df$start, special_chars_df$end)
    colnames(special_chars_df) <- c("Start", "End", "Name", "AA", "Anticodon", "Cellular_Localization", "Species", "ASCII_mod")
    return(special_chars_df)
  })
  do.call("rbind", bed_data_list)
}

bed_data <- generate_bed_data(yeast_modified)

# test write to bed file
output_file_path <- here("ref", "yeast_tRNA_modifications.bed")
write.table(bed_data, output_file_path, sep = "\t", quote = FALSE, row.names = FALSE, col.names = FALSE)

```

make the remaining bed files


make the remaining bed files, then let's look at how many unique characters are present in each species


Coming back to this 1/4, realizing there are NAs in the desired_species_data that contains the stuff we're definitely sequencing
```{r}
desired_species_data %>%
  filter(!is.na(Species))
```
