---
title: "phage"
author: "Laura White"
date: "2024-02-18"
output: html_document
---

```{r global options, include = FALSE}
knitr::opts_chunk$set(message=FALSE)
library(tidyverse)
library(colorblindr)
library(cowplot)
library(here)
here::i_am("Rmd/phage.Rmd")
```
We want to know if our E. coli T4  phage tRNA aligning reads are genuine.
Going to take the same basic approach as looking at mitochondrial reads.
```{r}
read_tRNA <- function(x, host, phage, time) {
  df <- readr::read_tsv(x, col_names = c("chrom", "start", "pos", "score"), show_col_types = FALSE)
  
  df <- df %>%
    dplyr::mutate(host = host, phage = phage, time = time)
  
  return(df)
}

# E. coli with and without T4 infection aligned to host + phage reference
read_tRNA(here("bedgraphs/Ecoli004_20240105_1224_MN31004_FAX73799_1fe84761.rna004_130bps_sup@v5.0.0.T4ref.bwa_full_length.bam.fl.bg"), "ecoli", "t4", "0_min") -> T4_uninfected
read_tRNA(here("bedgraphs/T4infectedecoli_20240202_1332_P2S-01618-A_PAU05281_fd3805fc.rna004_130bps_sup@v5.0.0.T4ref.bwa_full_length.bam.fl.bg"), "ecoli", "t4", "60_min") -> T4_infected

combined <- list(T4_uninfected, T4_infected)

bind_rows(combined) %>%
  mutate(pos = as.numeric(pos)) %>%
  mutate(score = as.numeric(score)) %>%
  separate(chrom, c("origin", "tRNA", "AA", "anticodon", "family", "isotype")) %>%
  select(-tRNA) %>%
  mutate(anticodon = str_replace_all(anticodon, "U", "T")) -> tRNAcov # u and t inconsistency across refs


rm(T4_uninfected, T4_infectedcombined)
```

```{r}
tRNAcov %>%
  group_by(origin, AA, anticodon, family, isotype, host, phage, time) %>%
  slice_max(score, with_ties = F) %>%
  ungroup() %>%
  unite(tRNA, origin, AA, anticodon, sep = "-") %>%
  select(-pos) -> rollup
```
The code below adds up max scores for family & gene number info so that we have summary scores for each isodecoder.
```{r}
rollup %>%
  separate(tRNA, c("origin", "AA", "anticodon"), extra = "drop") %>%
  group_by(origin, AA, anticodon, host, phage, time) %>%
  summarize(score = sum(score)) %>%
  unite(tRNA, origin, AA, anticodon, sep = "-") %>%
  ungroup() %>%
  group_by(host, phage, time) %>%
  mutate(total_score = sum(score)) %>%
  mutate(CPM = score/total_score*1000000) -> isodecoder_scores
```
Look at T4
```{r}
isodecoder_scores %>%
  filter(phage == "t4") %>%
  select(-phage, -host) %>%
  ungroup() %>%
  separate(tRNA, c("origin", "tRNA"), extra = "merge") -> t4scores

# Split the data into two parts based on time for easier manipulation
t4scores_0_min <- t4scores %>%
  filter(time == "0_min")

t4scores_60_min <- t4scores %>%
  filter(time == "60_min")

# Calculate fold change
t4scores_fold_change <- t4scores_60_min %>%
  left_join(t4scores_0_min %>% select(-time), by = c("tRNA", "origin"), suffix = c("_60_min", "_0_min")) %>%
  mutate(fold_change = CPM_60_min / CPM_0_min)

# Select relevant columns and adjust names as necessary
t4scores_fold_change <- t4scores_fold_change %>%
  select(tRNA, origin, fold_change)

ggplot(t4scores_fold_change, aes(x = fct_reorder(tRNA, fold_change), y = fold_change)) +
  geom_bar(stat = "identity", position = position_dodge(width = 0.7), width = 0.6) +
  geom_hline(yintercept = 1, color = "blue", linetype = "dashed", size = .5, alpha = .5) +
  scale_fill_OkabeIto(order = c(2, 1)) +
  theme_cowplot() +
  theme(axis.text.x = element_text(angle = 90, hjust = 0.5, vjust = 0.5)) +
  labs(x = "tRNA", y = "Fold change", title = "Change in host & phage tRNA expression on \n T4 phage infection of E. coli") +
  facet_wrap(~origin, ncol = 1, scales = "free_y")
```
That's a nice way to look at relationships between host and phage tRNAs of the same isodecoder family. 
But alternatively, let's keep the origin in the tRNA name field, so that we can color by origin and put them all on the same plot.
Also, calculate log2 fold change correctly down here.
```{r}

# Prepare and calculate the fold change
isodecoder_fold_change <- isodecoder_scores %>%
  select(-score, -total_score) %>%
  # Separate origin from the tRNA identifier more accurately
  mutate(origin = ifelse(grepl("^phage-", tRNA), "phage", "host"),
         tRNA = gsub("^(phage-|host-)", "", tRNA)) %>%
  pivot_wider(names_from = time, values_from = CPM) %>%
  drop_na(`60_min`, `0_min`) %>%
  # Calculate log2 fold change correctly without adding 1
  mutate(fold_change = log2(`60_min` / `0_min`),
         combined = paste(origin, tRNA, sep = "-"))  # Combine for display

# Plot with corrected data and axis labels
ggplot(isodecoder_fold_change, aes(y = fold_change, x = fct_reorder(combined, fold_change), fill = origin)) +
  geom_col(width = 0.6) +
  scale_fill_manual(values = c("host" = "grey", "phage" = "#009E73")) +  # Custom colors
  theme_cowplot() +
  scale_x_discrete(labels = function(x) gsub("^(host-|phage-)", "", x)) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  labs(y = "Log2 Fold Change", x = "", title = "Change in tRNA Abundance on T4 Infection of E. coli") +
  theme(legend.position = "top")

```

