---
title: "Read Lengths"
author: "Laura White"
date: "2023-12-21"
output: html_document
---

```{r}
knitr::opts_chunk$set(echo = F)
library(tidyverse)
library(cowplot)
library(colorblindr)
library(data.table)
library(here)

here::i_am("rmd/read_lengths.Rmd")
```
I've extracted read lengths for the aligned reads and for all reads for each sample x basecalling model. Let's take a look.
```{r}
allreads <- read_csv(here("fastq_read_lengths.txt"))

mappedreads <- read_csv(here("bam_aligned_read_lengths.txt"))
```
Make life easier by making data tables
```{r}
# Convert tibble to data.table
setDT(allreads)

# Extract Species from the beginning of the Sample name
allreads[, Species := fifelse(grepl("^Drosophila", Sample), "D_melanogaster",
                              fifelse(grepl("^(HumanFibroblast|Trub1doxH295R|Trub1untrH295R)", Sample), "H_sapiens",
                              fifelse(grepl("^(Pus4yeast|WTyeast)", Sample), "S_cerevisiae",
                              fifelse(grepl("^Tetrahymena", Sample), "T_thermophila",
                              fifelse(grepl("^Zebrafish", Sample), "D_rerio", NA_character_)))))]
                              
# Keep the last portion of the Sample name (from ".rna" on)
allreads[, BCChem := sub(".*\\.rna", "rna", Sample)]

# remove extraneous text in 'BCChem' column
allreads[, BCChem := sub("_\\d+bps|@v[0-9.]+", "", BCChem, perl = TRUE)]
allreads[, BCChem := sub("@.*$", "", BCChem)]

# same thing for mapped read tibble
setDT(mappedreads)

# Extract Species from the beginning of the Sample name
mappedreads[, Species := fifelse(grepl("^Drosophila", Sample), "D_melanogaster",
                              fifelse(grepl("^(HumanFibroblast|Trub1doxH295R|Trub1untrH295R)", Sample), "H_sapiens",
                              fifelse(grepl("^(Pus4yeast|WTyeast)", Sample), "S_cerevisiae",
                              fifelse(grepl("^Tetrahymena", Sample), "T_thermophila",
                              fifelse(grepl("^Zebrafish", Sample), "D_rerio", NA_character_)))))]
                              
# Keep the last portion of the Sample name (from ".rna" on)
mappedreads[, BCChem := sub(".*\\.rna", "rna", Sample)]

# remove extraneous text in 'BCChem' column
mappedreads[, BCChem := sub("_\\d+bps|@v[0-9.]+", "", BCChem, perl = TRUE)]
mappedreads[, BCChem := sub("bwa$", "", BCChem)]
mappedreads[, BCChem := sub("@.*$", "", BCChem)]

```


Basic exploratory plotting
```{r}
allreads %>%
  filter(!grepl("^(Trub1doxH295R|Trub1untrH295R|Pus4yeast)", Sample)) %>% # wildtypes only
  ggplot(., aes(x = ReadLength)) +
  geom_histogram(bins = 50, fill = "#0072B2", alpha = 1) +  # Adjust bin size as needed
  facet_grid(Species ~ BCChem, scales = "free_y") +
  labs(title = "Length distributions, all reads",
       x = "Read length",
       y = "Count") +
  xlim(c(0, 400)) + 
  theme_cowplot() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))

mappedreads %>%
  filter(!grepl("^(Trub1doxH295R|Trub1untrH295R|Pus4yeast)", Sample)) %>% # wildtypes only
  ggplot(., aes(x = AlignedReadLength)) +
  geom_histogram(bins = 50, fill = "#0072B2", alpha = 1) +  # Adjust bin size as needed
  facet_grid(Species ~ BCChem, scales = "free_y") +
  labs(title = "Length distributions, mapped reads",
       x = "Read length",
       y = "Count") +
  xlim(c(0, 400)) + 
  theme_cowplot() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))
```

