---
title: "tRNA abundance exploration"
author: "Laura White"
date: "2023-12-22"
output: html_document
---
```{r global options, include = FALSE}
knitr::opts_chunk$set(message=FALSE)
library(tidyverse)
library(colorblindr)
library(cowplot)
library(ggrepel)
library(pals)
library(forcats)
library(here)
here::i_am("Rmd/tRNAabundance_zebrafish.Rmd")
```

```{r}
read_tRNA <- function(x, species, chemistry, model, nt_threshold) {
  df <- readr::read_tsv(x, col_names = c("chrom", "start", "pos", "score", "nt_threshold"), show_col_types = FALSE)
  
  df <- df %>%
    dplyr::mutate(species = species, chemistry = chemistry, model = model, nt_threshold=nt_threshold) # easiest is add these by hand for now
  
  return(df)
}

read_tRNA(here("rebasecalling/bedgraphs/allreads/Zebrafish002_20231207_1528_P2S-00519-A_PAS44221_7e583c4c.rna002_70bps_fast@v3.bg"), "D_rerio", "RNA002", "fast", "none") -> fast2
read_tRNA(here("rebasecalling/bedgraphs/allreads/Zebrafish002_20231207_1528_P2S-00519-A_PAS44221_7e583c4c.rna002_70bps_hac@v3.bg"), "D_rerio", "RNA002", "hac", "none") -> hac2
read_tRNA(here("rebasecalling/bedgraphs/allreads/Zebrafish004_20231208_0919_MN31004_FAX73799_f3dea9f8.rna004_130bps_fast@v3.0.1.bg"), "D_rerio", "RNA004", "fast", "none") -> fast4
read_tRNA(here("rebasecalling/bedgraphs/allreads/Zebrafish004_20231208_0919_MN31004_FAX73799_f3dea9f8.rna004_130bps_hac@v3.0.1.bg"), "D_rerio", "RNA004", "hac", "none") -> hac4
read_tRNA(here("rebasecalling/bedgraphs/allreads/Zebrafish004_20231208_0919_MN31004_FAX73799_f3dea9f8.rna004_130bps_sup@v3.0.1.bg"), "D_rerio", "RNA004", "sup", "none") -> sup4
read_tRNA(here("rebasecalling/bedgraphs/70ntplus_reads/Zebrafish002_20231207_1528_P2S-00519-A_PAS44221_7e583c4c.rna002_70bps_fast@v3.bg"), "D_rerio", "RNA002", "fast", "70") -> fast2_70
read_tRNA(here("rebasecalling/bedgraphs/70ntplus_reads/Zebrafish002_20231207_1528_P2S-00519-A_PAS44221_7e583c4c.rna002_70bps_hac@v3.bg"), "D_rerio", "RNA002", "hac", "70") -> hac2_70
read_tRNA(here("rebasecalling/bedgraphs/70ntplus_reads/Zebrafish004_20231208_0919_MN31004_FAX73799_f3dea9f8.rna004_130bps_fast@v3.0.1.bg"), "D_rerio", "RNA004", "fast", "70") -> fast4_70
read_tRNA(here("rebasecalling/bedgraphs/70ntplus_reads/Zebrafish004_20231208_0919_MN31004_FAX73799_f3dea9f8.rna004_130bps_hac@v3.0.1.bg"), "D_rerio", "RNA004", "hac", "70") -> hac4_70
read_tRNA(here("rebasecalling/bedgraphs/70ntplus_reads/Zebrafish004_20231208_0919_MN31004_FAX73799_f3dea9f8.rna004_130bps_sup@v3.0.1.bg"), "D_rerio", "RNA004", "sup", "70") -> sup4_70

combined <- list(fast2, hac2, fast4, hac4, sup4, fast2_70, hac2_70, fast4_70, hac4_70, sup4_70)

bind_rows(combined) %>%
  mutate(pos = as.numeric(pos)) %>%
  mutate(score = as.numeric(score)) %>%
  separate(chrom, c("tRNA", "AA", "anticodon", "family", "isotype")) %>%
  select(-tRNA) -> tRNAcov

rm(fast2, hac2, fast4, hac4, sup4, fast2_70, hac2_70, fast4_70, hac4_70, sup4_70, combined)
```

The above gives us a tidy tibble with coverage x position for each tRNA in each sample.
I then make the assumption that the maximum coverage score for each gene is a better proxy coverage value than (for example) the mean coverage on that reference, because per-nucleotide coverage is being depressed by Nanopore error rates and mismatches/indels due to tRNA modifications.
```{r}
tRNAcov %>%
  group_by(species, chemistry, model, AA, anticodon, family, isotype, nt_threshold) %>%
  slice_max(score, with_ties = F) %>%
  ungroup() %>%
  unite(tRNA, AA, anticodon, sep = "-") %>%
  select(-pos) -> rollup
```

The code below adds up max scores for family & gene number info so that we have summary scores for each isodecoder.
```{r}
rollup %>%
  separate(tRNA, c("AA", "anticodon"), extra = "drop") %>%
  group_by(AA, anticodon, species, chemistry, model, nt_threshold) %>%
  summarize(score = sum(score)) %>%
  unite(tRNA, AA, anticodon, sep = "-") %>%
  ungroup() %>%
  group_by(species, chemistry, model, nt_threshold) %>%
  mutate(total_score = sum(score)) %>%
  mutate(CPM = score/total_score*1000000) -> isodecoder_scores
```

Plot these summed scores for all isodecoders, fast basecallling
```{r}
# Reshape the data
cpm_comparison <- isodecoder_scores %>%
  filter(model == "fast" & nt_threshold == "70") %>%  # select model and filter
  select(tRNA, species, chemistry, model, CPM) %>%
  pivot_wider(names_from = chemistry, values_from = CPM, names_prefix = "CPM_") %>%
  separate(tRNA, into = c("isoacceptor", "tRNA"), sep = "-", remove = FALSE)

# calculate pearson's rho
cor_coefficient <- cor(cpm_comparison$CPM_RNA002, cpm_comparison$CPM_RNA004, method = "pearson")

# specify color palette for isoacceptor labeling
tol_rainbow_colors <- pals::tol.rainbow(23)

ggplot(cpm_comparison, aes(x = CPM_RNA002, y = CPM_RNA004, color = isoacceptor)) +
  geom_point() +
  scale_color_manual(values = tol_rainbow_colors) +
  geom_text_repel(aes(label = tRNA), size = 3, color = "black") +
  scale_x_log10() +
  scale_y_log10() +
  labs(title = expression(italic("D. rerio")*" normalized read counts"),
       x = "RNA002",
       y = "RNA004",
       subtitle = sprintf("Pearson's Ï: %.2f", cor_coefficient)) +
  theme_cowplot() +
  theme(aspect.ratio = 1,
        plot.margin = margin(5.5, 5.5, 15.5, 5.5, "points"), # Increase bottom margin
        plot.subtitle = element_text(hjust = 0.5, size = 12, face = "italic", color = "black")) +
  coord_fixed(ratio = 1)
```
Now I want to do the same thing, but compare a best case basecalling for RNA002 vs. the same for RNA004 (so different chemistries, hac vs sup).
```{r}
isodecoder_scores %>%
  ungroup() %>%
  mutate(chemistry_model = paste(chemistry, model, sep = "_")) %>%
  select(species, tRNA, CPM, chemistry_model, nt_threshold) %>%
  filter(chemistry_model %in% c("RNA002_hac", "RNA004_sup") & nt_threshold == "70") -> subset_scores

# Reshape the data
cpm_comparison <- subset_scores %>%
  pivot_wider(names_from = chemistry_model, values_from = CPM) %>%
  na.omit()  # Remove rows with NA values

cor_coefficient <- cor(cpm_comparison$RNA002_hac, cpm_comparison$RNA004_sup, method = "pearson")

# Create the plot with regression line and repelling labels
ggplot(cpm_comparison, aes(x = RNA002_hac, y = RNA004_sup)) +
  geom_point() +
  geom_text_repel(aes(label = tRNA), size = 3) +
  scale_x_log10() +
  scale_y_log10() +
  labs(title = "Counts per million reads, best available basecalling (D. rerio)",
       x = "RNA002",
       y = "RNA004") +
  theme_cowplot() +
  coord_fixed(ratio = 1) +
  annotate("text", x = Inf, y = Inf, label = sprintf("Pearson's r: %.2f", cor_coefficient),
           hjust = 2.5, vjust = 2, size = 4, color = "#0072B2")

```

