---
title: "tRNA abundance exploration"
author: "Laura White"
date: "2023-12-22"
output: html_document
---
```{r global options, include = FALSE}
knitr::opts_chunk$set(message=FALSE)
library(tidyverse)
library(colorblindr)
library(cowplot)
library(ggrepel)
library(forcats)
library(here)
here::i_am("Rmd/tRNAabundance_human.Rmd")
```

```{r}
read_tRNA <- function(x, species, chemistry, model) {
  df <- readr::read_tsv(x, col_names = c("chrom", "start", "pos", "score"))
  
  df <- df %>%
    dplyr::mutate(species = species, chemistry = chemistry, model = model) # easiest is add these by hand for now
  
  return(df)
}

read_tRNA(here("rebasecalling/bedgraphs/HumanFibroblast002_20231207_1455_P2S-00519-B_PAS29437_b7f6c4bd.rna002_70bps_fast@v3.bg"), "H_sapiens", "RNA002", "fast") -> fast2
read_tRNA(here("rebasecalling/bedgraphs/HumanFibroblast002_20231207_1455_P2S-00519-B_PAS29437_b7f6c4bd.rna002_70bps_hac@v3.bg"), "H_sapiens", "RNA002", "hac") -> hac2
read_tRNA(here("rebasecalling/bedgraphs/HumanFibroblast004_20231208_1351_MN42516_FAX71838_b8f7b990.rna004_130bps_fast@v3.0.1.bg"), "H_sapiens", "RNA004", "fast") -> fast4
read_tRNA(here("rebasecalling/bedgraphs/HumanFibroblast004_20231208_1351_MN42516_FAX71838_b8f7b990.rna004_130bps_hac@v3.0.1.bg"), "H_sapiens", "RNA004", "hac") -> hac4
read_tRNA(here("rebasecalling/bedgraphs/HumanFibroblast004_20231208_1351_MN42516_FAX71838_b8f7b990.rna004_130bps_sup@v3.0.1.bg"), "H_sapiens", "RNA004", "sup") -> sup4

combined <- list(fast2, hac2, fast4, hac4, sup4)

bind_rows(combined) %>%
  mutate(pos = as.numeric(pos)) %>%
  mutate(score = as.numeric(score)) %>%
  separate(chrom, c("tRNA", "AA", "anticodon", "family", "isotype")) %>%
  select(-tRNA) -> tRNAcov

rm(fast2, hac2, fast4, hac4, sup4, combined)
```

The above gives us a tidy tibble with coverage x position for each tRNA in each sample.
I then make the assumption that the maximum coverage score for each gene is a better proxy coverage value than (for example) the mean coverage on that reference, because per-nucleotide coverage is being depressed by Nanopore error rates and mismatches/indels due to tRNA modifications.
```{r}
tRNAcov %>%
  group_by(species, chemistry, model, AA, anticodon, family, isotype) %>%
  slice_max(score, with_ties = F) %>%
  ungroup() %>%
  unite(tRNA, AA, anticodon, sep = "-") %>%
  select(-pos) -> rollup
```

The code below adds up max scores for family & gene number info so that we have summary scores for each isodecoder.
```{r}
rollup %>%
  separate(tRNA, c("AA", "anticodon"), extra = "drop") %>%
  group_by(AA, anticodon, species, chemistry, model) %>%
  summarize(score = sum(score)) %>%
  unite(tRNA, AA, anticodon, sep = "-") %>%
  ungroup() %>%
  group_by(species, chemistry, model) %>%
  mutate(total_score = sum(score)) %>%
  mutate(CPM = score/total_score*1000000) -> isodecoder_scores
```

Plot these summed scores for all isodecoders, fast basecallling
```{r}
# Reshape the data
cpm_comparison <- isodecoder_scores %>%
  filter(model == "hac") %>%  # select model
  select(tRNA, species, chemistry, model, CPM) %>%
  pivot_wider(names_from = chemistry, values_from = CPM, names_prefix = "CPM_") %>%
  na.omit()  # Remove rows with NA values

# Create the plot with regression line and repelling labels
plot <- ggplot(cpm_comparison, aes(x = CPM_RNA002, y = CPM_RNA004)) +
  geom_point() +
  geom_text_repel(aes(label = tRNA), size = 3) +
  geom_smooth(method = "lm", se = FALSE, color = "#0072B2") +
  scale_x_log10() +
  scale_y_log10() +
  labs(title = "Counts per million reads, high accuracy basecalling (H. sapiens)",
       x = "RNA002",
       y = "RNA004") +
  theme_minimal() +
  coord_fixed(ratio = 1)

# Add R-squared value
model <- lm(CPM_RNA004 ~ CPM_RNA002, data = cpm_comparison)
r_squared <- summary(model)$r.squared
plot + annotate("text", x = Inf, y = Inf, label = paste("R² =", round(r_squared, 2)), hjust = 1, vjust = 1, position = "topright")

```
Now I want to do the same thing, but compare a best case basecalling for RNA002 vs. the same for RNA004 (so different chemistries, hac vs sup).
```{r}
isodecoder_scores %>%
  ungroup() %>%
  mutate(chemistry_model = paste(chemistry, model, sep = "_")) %>%
  select(species, tRNA, CPM, chemistry_model) %>%
  filter(chemistry_model %in% c("RNA002_hac", "RNA004_sup")) -> subset_scores

# Reshape the data
cpm_comparison <- subset_scores %>%
  pivot_wider(names_from = chemistry_model, values_from = CPM) %>%
  na.omit()  # Remove rows with NA values

# Create the plot with regression line and repelling labels
plot <- ggplot(cpm_comparison, aes(x = RNA002_hac, y = RNA004_sup)) +
  geom_point() +
  geom_text_repel(aes(label = tRNA), size = 3) +
  geom_smooth(method = "lm", se = FALSE, color = "#0072B2") +
  scale_x_log10() +
  scale_y_log10() +
  labs(title = "Counts per million reads, best available basecalling \n (H. sapiens)",
       x = "RNA002",
       y = "RNA004") +
  theme_minimal() +
  coord_fixed(ratio = 1)

# Add R-squared value
model <- lm( RNA004_sup ~ RNA002_hac, data = cpm_comparison)
r_squared <- summary(model)$r.squared
plot + annotate("text", x = Inf, y = Inf, label = paste("R² =", round(r_squared, 2)), hjust = 5, vjust = 1, position = "topright")

```

